### DISCUSSION TITLE
refine configuration provisioning

### DISCUSSION THREAD

**Discussion by hellt** [srl-labs/containerlab#487] [Ideas]
Hi @kellerza @karimra 
I merged the #383. Sorry it took so long, @kellerza, I wish I could get to it sooner... Thank you for stepping in and delivering the feature implementation of the config generation and provisioning. It was a lot of work, and you made a very important first step!

I propose to open this forum to discuss how do we take it from here, because although it is a working solution, it has quite a few things which are somewhat opinionated and not generic.

Let me summarize here what I did on top of the #383 after the last commit of @kellerza:

## 1. the way to provide variables for templation

I followed the original idea of providing variables for nodes and links to not introduce any breaking change. But we need to discuss this, as we may have a stronger design here. @karimra that is the issue we discussed the other day.

Originally the `labels` attribute of nodes and links was used to keep variables that are used in config templation.
I changed that to the following:

1. Node variables are defined with `.topology.nodes.<node>.config.vars`. This is a `map[string]string` structure which allows flat variables pairs in the form of `var-name:var-value`.
2. link variables are provided with `.topology.links[].vars`

That change required me to change the original `.topology.nodes.<node>.config` to ``.topology.nodes.<node>.startup-config` to free up the `config` key.

The `.topology.nodes.<node>.config` is of type [*ConfigDispatcher](https://github.com/srl-labs/containerlab/blob/master/types/node_definition.go#L9). It is not a simple map for a single reason that it was envisioned to grow and incorporate transport selection, transport options, template names to use.

Consequently, I had to create a few methods for Topology and NodeDefinition struct, namely [GetConfigDispatcher](https://github.com/srl-labs/containerlab/blob/master/types/node_definition.go#L62)

An example of the variables defined in clab file:

```yaml
name: srlcfg

topology:
  nodes:
    srl1:
      kind: srl
      image: registry.srlinux.dev/pub/srlinux:21.3.1-410
      license: license.key
      config:
        vars:
          systemip: 10.0.0.1/32

    srl2:
      kind: srl
      image: registry.srlinux.dev/pub/srlinux:21.3.1-410
      license: license.key
      config:
        vars:
          systemip: 10.0.0.2/32

  links:
    - endpoints: ["srl1:e1-1", "srl2:e1-1"]
      vars:
        port: ethernet-1/1, ethernet-1/1
```

## 2 providing templates to use for templation

The templates that will be rendered on a **per-node** basis are provided via two flags of the `config` subcommand

* `--template-path` - recieves a list of paths where the actual template files will be looked at
* `--template-list` - takes a list of names of the templates files that will be searched in paths provided with the flag above

The `--template-list` requires a bit of an explanation. The pattern that the templates of the current code are using is `<name>__<role>.tmpl`.

The `name` is a free formed string, whereas the `role` is a suffix that denotes to which role this template belongs to. The idea here is that in a multivendor lab or in a lab with multiple roles the devices would require different templates.
The `role` by default is assumed to be the `kind` of the node, but it is configurable.

Let's imagine we want to create a template that configures eBGP between SRL and SR OS, we could create the following files

* `bgp__srl.tmpl`
* `bgp__vr-sros.tmpl`

and then by doing `clab config -t <topo> --template-path <path> --template-list bgp` we are able to pick the right template per each node of the lab, based on the kind selector.

A very basic example of the template for interface provisioning on SRL can look like this:

```
{{ expect .systemip "ip" }}
{{ range .links }}
    {{ expect .port "^(ethernet-\\d+/|e\\d+-)\\d+$" }}
    {{ expect .name "string" }}
{{ end }}

/interface lo0 {
    admin-state enable
    subinterface 0 {
        ipv4 {
            address {{ .systemip }} {
            }
        }
        ipv6 {
            address ::c1ab:{{ ip .systemip }}/128 {
            }
        }
    }
}

/network-instance default {
    router-id {{ ip .systemip }}
    interface lo0.0 {
    }
}

/system lldp admin-state enable

/ network-instance mgmt {
    description "set from clab"
}

{{ range .links }}
/interface {{ .port }} {
    admin-state enable
}
{{ end }}
```

If, say, you have this file under `./private/iface__srl.tmpl`, then you can configure your SRL nodes of the lab shown above with the following command

```bash
# by omitting --template-list the code will take all files available in the search path
clab config -t srl-cfg-demo.clab.yml --tempalte-path ./private 
```

# Open points
1. I know that @karimra wanted to re-eval the way the vars are split between the nodes and links. Our ultimate goal is to be able to specify tree based vars, to avoid the limitations of a flat variables structure.

2. @kellerza  It would be interesting to see what warranted to create a wrapper around stdlib `text/template`. I am not a fan of using custom wrappers until necessary. I remember before you had a functions map that you supplied to `text/template`

3. regardless of 1) there are a few hard-coded assumptions in the code that does variables preparation for links. For example, it is expected that system ip is always provided in the vars - https://github.com/srl-labs/containerlab/blob/929abda141fbfc1985c9ccbee6ba51d9c1959f97/clab/config/utils.go#L75
I would rather assume that we should let users decide which variables need to be present. The code should only merge variables and pass them to the template

**Comment by kellerza** [comment]
🎉 !

# open points

## 2 template wrapper

This contains various helper [functions](https://pkg.go.dev/github.com/kellerza/template) (like *ip* & *expect* used in your example template above splitting&testing used in the sros template) and respective tests for these. I also overwrite standard text/template functions to enable them to be piped & at the same time keep them backward compatible to text/template.

It was another 800-1000 lines of code that is purely used to make text/template usable for this purpose, and adding some functions that one would expect coming from ansible/jinja. So during the review we decided to split it out

in the future functions for assigning IPs from pools can be added here as well

## 3 variable prep

This is quite important for links, else you have to explicitly specify an IP on each, but with some more thought I can probably move this to template functions as well

**Reply by karimra** [reply]
@kellerza which part of the custom template package cannot be done with the standard template package and adding custom functions to it ?
The external package can be a collection of custom functions, not a whole wrapper of the std package.
There is also a well known Go template functions package that augments the std template package: https://github.com/Masterminds/sprig

**Reply by kellerza** [reply]
The majority of the feedback so far is that people have used/has templates with Jinja syntax, so sprig based on underscore function are not ideal. They also don't overwrite standard functions. There are some other, but in all cases it will also need extending. Things like negative indexes is important for me, etc. The SROS example in the PR uses a lot more of these

They also do not have the use case for expect/default, to supply mandatory/optional parameters with validation.

The entire wrapper today only adds the functions, but my intention is that is can convert variables to not be flat, as in the following example, to provide a cleaner template syntax

Today:
```
{{ expect .systemip "ip" }}
ip {{ ip .systemip }}
mask {{ mask .systemip }}
```

can become
```
{{ expect .systemip "ip" }}
ip {{ .systemip.ip }}
mask {{ .systemip.mask }}
```

**Reply by karimra** [reply]
I'm not suggesting to use sprig specifically, I'm saying the std template package can be augmented with functions, 
there is no need to create a wrapper package. 
You can add multiple function maps as well, your own custom functions as well as sprig.

If the wrapper adds only functions why have it?
I can see it adds some restrictions, like assuming the variables are always map[string]interface{}, and not a list.

if we go this way, we are creating a containerlab specific templating syntax, I don't think this is a good idea.
There was multiple attempts at porting jinja2 or twig to Go, or creating new templating syntax, they all failed or stopped being supported. The only safe bet is to stick with Go template.

**Reply by kellerza** [reply]
I don't have any problem if we just load the template functions. We wont be able to expand the `expect` function in that case, but we can live without it.

It might also limit autogen functions (like IP/number pools) if you don't control the entire template rendering, but lets see when we get there

**Reply by hellt** [reply]
I am in favor of using the stdlib template engine over the wrapper and using functions as the means of extension.
The reason is to keep the template design language as used in Go and not deviate towards a bespoke solution. Having custom functions should get us pretty far in making the template feature-rich, but at the same time the overall language is still Go and users will be familiar with it inherently.

I wouldn't look too much on jinja at the same time, it is defo there and many people know it, but so does Go templates getting more and more popular.

So I propose we either
1. move the template functions in a separate package if you wish, but without wrapping the template engine
2. move the functions to clab

going with 1) can be easier if you plan to reuse those functions in other packages (which I see can be a valuable thing)

**Reply by kellerza** [reply]
Indeed (1) can be re-used. This is already possible today, by using [template.Funcs](https://pkg.go.dev/github.com/kellerza/template#pkg-variables). All template functions are public in order to have them visible with godocs

To unwrap the template package is a clab only change. I'll move some of the wrapper code from the package to clab in a future PR

**Comment by karimra** [comment]
## 1. the way to provide variables for templation

Having two sections to define variables can cause some issues, 
* What happens if the same key is defined under `config.vars` and `links.[].vars` ? which one takes precedence ?
* You will also have 2 distinct parts to keep track of for each node.
* What distinguishes a node variable vs a link variable? looks like an unnecessary subjective decision.

I think it's better to keep variables in one place. The link's labels can be used to influence how an interface is configured, e.g:
( not proper template syntax )

```gotmpl
{{ range $link := .Links }}
  {{ if hasLabel $link "label1" }}
    CfgA
  {{ else if hasLabel $link "label2" }}
    CfgB
  {{ end}}
{{ end}}
```

## 2. providing templates to use for templation

The templates can be a global variable, a path that is recursively parsed to load all the templates.
the loaded templated are then matched with the nodes using the file name, according to the same proposed schema <name>__<kind>.tmpl

## Open points

1. same as what is written in 1. above, flat variables structure can be avoided with map[string]interface{}. We just have to decide how deep a merge should go, how to merge lists and how to handle duplicated items in a list.

**Reply by kellerza** [reply]
## 1. providing variables
Not sure why there would ever be a need to merge links & nodes. There will be things that are different between nodes & links, but the same var/concept

Links should get a `node` and `node_far`, which are both struct and contain the entire local & far end node structure. This also removes some of the assumption in the Link var preparation functions. I still wanted to add this, but didnt want to create another moving target for this base work.

The template wrapper already takes `map[string]interface{}` for this purpose, not to have flat variables. Whether they come from nodes, mutations due to expect functions, or simply specified as nested itemsin the yaml (on both node & link level).

No complicated merging should be required with this approach ;-)

## 2 providing template.
Not sure I follow the global variable statement, but this is also another function of the template wrapper. It recursively searches for templates in the list of folders. So you can overwrite some templates with your own. or use a mix of your own and base templates.

**Reply by karimra** [reply]
1. So you are assuming that a template uses either the variables under node or the ones under links, but not both ?
the merging will be required for variables specified on kind and defaults levels as well.

2. the std lib does just that, there is really no need to create a wrapper for this

**Reply by kellerza** [reply]
1. Each node has an array of links. Each links also points to `node` and `node_far`. The `node` can be seen as a circular ref to the node containing the link.
Agree defaults&kinds should be merged. Maybe use mergo? (just saw it is used by swig)
2. In that case my bad, I missed this. Is it similar to [this](https://github.com/kellerza/template/blob/main/template.go#L59)?

**Reply by hellt** [reply]
1. for the merging of variables defined on `defaults->kind->node` can't we use a json merge method? i.e. we marshal the `config.vars` to json, and then use a json patch function on json data to get the resulting json that then can be unmarshalled to `map[string]interface]`?

https://github.com/evanphx/json-patch#combine-merge-patches

this should let us merge with indefinite depth I assume

**Reply by kellerza** [reply]
Could work, as long as we do this before we create the circular node->link->node. So before we add `node` and `node_far` to each link.

**Reply by hellt** [reply]
as to the merge of nodes and links that @karimra referred above I suppose he was talking about the way the vars are split between the nodes and the link, for example:

```yaml
topology:
  nodes:
    srl1:
      kind: srl
      config:
        vars:
          systemip: 10.0.0.1/32

    srl2:
      kind: srl
      config:
        vars:
          systemip: 10.0.0.2/32

  links:
    - endpoints: ["srl1:e1-1", "srl2:e1-1"]
      vars:
        port: ethernet-1/1, ethernet-1/1
```

The question is can we not use the vars under the links and put them all under the node?

**Reply by kellerza** [reply]
> The question is can we not use the vars under the links and put them all under the node?

Then we are back at splitting the definition and vars: How many links, what if you change the order of the links, add/remove a link. If its not together you lose that context

(From you example we will remove the logical ports with #377, but there are many other reasons to tie vars to links)

**Reply by kellerza** [reply]
Except if we can numbering the links/provide a unique link identifier that is also backward compatible. 

It will still be visually split, but at least we'll have a way to reference the correct link

**Reply by hellt** [reply]
I am not sure that added logic for merging links and the corresponding nodes warrant the need for that.
Meaning that I would like to explore the different, more straightforward approach where vars are defined under each node only (and kinds/defaults).
Yes, you, as a template designer, will need to define the link variables, but links can be tied in with interface name.

For example

```yaml
topology:
  nodes:
    srl1:
      kind: srl
      config:
        vars:
          systemip: 10.0.0.1/32
          links:
            - name: ethernet-1/1
              other-var: value
```

**Reply by kellerza** [reply]
How do you get the far end other-var? (Only via building the state of the links)

this will be super complicated and you just split up a link. An example is something like isis-instace-id should it now live on side a, side b, or duplicated across two branches?
Even conflict resolution here becomes more complex

**Reply by hellt** [reply]
I assume that per-node vars approach means that you don't have this tie in with links. Meaning that the `other-var` of a link should be specified on both nodes variables. @karimra do you have the same in mind?

Maybe what will help here is to discuss a particular real example. @kellerza can you paste here a piece of clab file where link vars provide the simplification that can't happen with node-only vars?
You mentioned `isis-instance-id`, can you put some context with the template to demonstrate how link-scoped vars achieve some benefits?

**Reply by kellerza** [reply]
The whole PR was done to illustrate this point. 

If you look at the result, start with only the variables (try adding: -c 1 -d) with the simple topo files in lab-examples/vr05 this will help.

Then, from a template perspective, just try to answer these questions:

- how do you get the interface name? (Typically refer to the remote side)
- How do you indicate a bfd target for a link/lsp. It is not the local IP, it is the far end IP. Might even need the far end system ip (hence the need for node_far.systemip)
- Instance-id. Build 2 igp areas. The instance Id must match, currently specifying it once is enough. 
- specifying vlan once is enough. Vlan must also match. The lab example shows a vlan,vlan as well (specify it explicitly, but right next to each other)

For all cases above, the existing implementation either does not specify anything in the topo file, or at worst does it once. 

With your suggestion it seems like everything needs to be explicitly specified. And probably duplicated across vastly different yaml locations, or is this not the case?

**Reply by karimra** [reply]
I think I get what you did with the PR, but I have to say it is a very specific use case.
Knowing which interfaces belong to the same link can be done in other ways that are a bit more flexible: 
- Using labels, and/or
- Passing the whole topology to the template

The logic on how the vars will be used will be moved to the template in those case.

**Reply by kellerza** [reply]
Not sure why the mechanism is specific? 

With this approach you supply the template designer with variables they may need. 

If they don't like the default that is something else, nothing stops them from being very verbose and provide their own more specifics (like name, link IPs etc). The code allows for that, today! 
(Having variables assigned from a pool is a logical extension of this)

I fail to see how passing this all to the template designer will help our case:
- The problem with Jinja template is that you can embed a lot of logic and in many cases these become unreadable.
- So Go templates solve this by being very basic functions and they prefer you roll a lot of logic, formatting etc, into a go function, that can be tested externally and it results in simpler templates.
- but, In Containerlab, as an application user, you do not have the option to add your own go function as a template designer, you will be 100% reliant on what the implementation provides.

So if we remove the entire concept of having the `*_far` end variables passed into the template, I'm not sure this is workable:
Go templates don't have a global context, so passing the whole topology will be really nice, until you iterate over the links with range, then you don't have access to it anymore. 😦  The labels approach might suffer the same fate, as I understand label would be to lookup something like a far end. Much better just to pass the far end, as part of the link, to the template and the template designer can use it/ignore it as they see fit

Now I've seen some strange hacks to bind context to closures before you load any template, but all the arguments you used against the harmless template wrapper applies to this and I'm sure we can add a couple more (to date it seems Go devs don't like this approach either)

You can probably say the templates are specific and I'll agree they are slightly opinionated, but this is just an example to illustrate why you need some of these concepts in the template machinery.

**Reply by karimra** [reply]
It is specific because of the variables added to links, those variables are tightly linked to the links and the nodes.

I never said there is a need for a global context for the templates, that can introduce race conditions when executing the templates.

I don't get this statement:
```
..., until you iterate over the links with range, then you don't have access to it anymore
```
what do you mean by don't have access to it anymore ?

the labels are simple variables, tags if you will. The user is free to set them as they want. how they are used is up to the user and is done in the template.

**Reply by kellerza** [reply]
> what do you mean by don't have access to it anymore ?

inside a `{{ range .x }}` you loose the original `.` context and only have access to that member of `.x`, represented as `.`.

So you need more than just simple labels and without `node` and `node_far` info on each link, which is the bare minimum, this cannot work

**Reply by kellerza** [reply]
I don't have objections that you change a working implementation. All I ask is that you keep it working.

**Reply by kellerza** [reply]
And ideally not bloat the info you need to provide in the yaml file & the template in the process, but this is not a critical requirement for usability, just a nice to have

**Reply by hellt** [reply]
@kellerza @karimra  It seems we agreed so far on the following action points. Do you have any comments to any of them?

1. Remove the wrapper arounds the `text/template` and use solely custom functions. With the reason being to not introduce unnecessary shadowing of the stdlib functions

2. Remove the hard requirements on certain variables to be present. The goal is to not enforce an opinion on variable names or their availability. The templation framework should be agnostic to that, leaving such decisions to be made by a user.

3. Remove the requirement to have `vars` to be defined in the `link` section. Users can still have vars defined under links, and thus leverage the "added-logic" functions you have already developed. But this should not be a hard requirement.
If a user wants to provide bare variables under the node only, or even load them from a separate files this should not stop them from doing that.
Implementing that, will allow for many templation approaches to coexist. Users will be the one who decide which one they want to use, and I find this quite important.

4. Introduce the free-formed variables structure (`map[string]interface{}`) instead of a flat list of key-val pairs.

It is never easy to work out a common opinion, given the views we all have are subjective and are based on our experience both with Go templates as well as with other frameworks. So I hope no-one has any hard feelings about others having other opinions ;) Thus I feel that the above points make it possible to unlock the future development while also making it more "flexible" in the ways the templation can be implemented.

**Reply by kellerza** [reply]
No hard feelings, but so far you are quick to propose removing functionality, without in most cases trying to understand why it is what it is, or having workable alternatives. (not duplication!)

2. Any hard requirement can be made optional (Most of them are already like this, its probably only systemip left). 
  From the start we agreed that tying the topology to the logical vars brings benefits & I've stated my case why I see this as critical. If we choose not to leverage the topology, I honestly don't see this as something I will use.

3. There is no hard requirement today. 
  There is also no way to logically configure even basic ports without it though. In SROS for sure, but likely anything other than SRL. Actually, even in SRL, you also need the topology info under the links

4. This was the intention from the start and we are only here because 'labels' was duplicated as 'vars' 🤷

**Reply by hellt** [reply]
I am quick not to remove the functionality, but quick to remove the lock-in on the options. In the current PR there is this lock-in that forces a certain way of writing templates, and only this way. By making this not to be locked-in we avoid finding ourselves in a situation when the next user comes in and says something along these lines

> If there is no way to provide vars in a separate file, I honestly don't see this as something I will use.

In other words, the functionality stays, but it must be optional.

2. Yes, the benefits of tying in topo are still there. All I am voting for is not making this the *only* way to generate configurations.

3. There is always a way to configure anything using only vars on the nodes. One very simple way is to specifically say what you want to configure on which node. Yes, there is a repetition in vars with that way, but it is simple and I bet it will be used by some people as well.

**Reply by kellerza** [reply]
I’ve written some of [this](https://github.com/home-assistant/core/blob/dev/homeassistant/util/yaml/loader.py) including can become a rabbit hole 😆 

ok, will make it optional and on top I’ll try move some of this to template functions (as long as these are passed as refs it should be easy to mutate the link maps in the template)

**Reply by karimra** [reply]
```
inside a {{ range .x }} you loose the original . context and only have access to that member of .x, represented as ..
```

Not true, there are multiple ways to access the outer scope:

1. `$` always refers to the variable passed to the template, you can always use `$.OuterScopeVar`
```
{{ range .links }}
 {{ $.OuterScopeVar }}
 {{ .IPAddr }}
{{ end }}
```
2. define a variable to range over, `.` will still refer to the main context:
``` 
{{ range $item := .links }}
 {{ .OuterScopeVar }}
 {{ $item.IPAddr }}
{{ end }}
```
you can even add an index, just like in Go
``` 
{{ range $index, $item := .links }}
 {{ .OuterScopeVar }}
 {{ $item.IPAddr }} has index {{ $index }}
{{ end }}
```

3.  create your own variables:
```
{{ $save := . }}
{{ range .links }}
 {{ $save.OuterScopeVar }}
 {{ .IPAddr }}
{{ end }}
```

Go template syntax/functions are as rich as jinja2, it just hasn't been around as long.
If projects like Docker, Helm, Hugo, K8s,... can live with them, I think clab can as well.

I hated Go template syntax when I discovered it, why create something new while we have jinja2 ?
The Go devs have their reasons, that are probably beyond me. 
So I just went and learned it, it turns out all the tools you need are already there.

Don't get me wrong, I'm not "against" what you added, I just think it can be done in a simpler way, while keeping it close to the Go std lib. 
The goal is to have as many people contribute, if clab has its own template syntax/engine, that adds another barrier to entry.

**Reply by kellerza** [reply]
Thanks! Didnt’t know about `$.`, that solves local `node` under links, but there is no way for `node_far` as it is not part of the template to begin with

Maybe we can just call it `far` /  `remote`, although prefer far. But it needs to be there. Even if you want to add a nodes map to the toplevel, and it is just a lookup.

A toplevel nodes map will also allow you to create templates for any sort of mesh.

**Reply by karimra** [reply]
I don't get why you say it's not possible.
Given a node name and an interface name, the far node can be retrieved. That can be one of the functions added to the template. 

Can you explain why it's not possible?

**Reply by kellerza** [reply]
Well maybe you can do it if that template function uses global vars, but is this really a good design? (This is in my view similar to trying to bind some context to the template function)

I would much prefer a design where all vars needed for execution is passed to the template render function

When you render a node it does not have access to other nodes or the topo file. Only what you passed to it, which is the labels, links (+labels) and these `_far` end variables I refer to

**Reply by karimra** [reply]
There is no need to use global vars.

Together with the node, the whole topology can be passed to the template.
something like:

```go
type templateInput struct {
   Node     *types.NodeConfig   // the node for which the template is rendered, even if it can be considered redundant
   Topology *types.Topology
}
```

**Reply by kellerza** [reply]
So now we have gone full circle. 

So lets just keep on adding variables to the map we pass today. Not enforcing any specific keys to be present and leave it up to the template designer to use what they need

**Reply by karimra** [reply]
There is no enforcement of any key to be present, I'm not sure where you see this.

If the struct above is used as is, the variables are just accessed in a different way. `.Node.Config.Vars`
To the same struct, a `Vars map[string]interface{}` can be added, the vars will be access with `.Vars`
If you don't like the `.Vars` notation, add a function `vars` or `config` that returns `.Vars`

I'm not saying it should be this way, just saying that there are other ways to achieve the same things

**Comment by kellerza** [comment]
#495 addresses the issues of required `systemip` and all additional vars should now be optional. Also posted a set of slide on the Teams channel

@hellt you want me to work on # 4?
> Introduce the free-formed variables structure `map[string]interface{}` instead of a flat list of key-val pairs.

**Reply by hellt** [reply]
thanks @kellerza 
will have a look at #495

sure, if you can, please feel free to "Introduce the free-formed variables structure map[string]interface{} instead of a flat list of key-val pairs."
